src/intrinsics/ecma262/Object.js

```diff
@@ -41,6 +41,92 @@ import { Create, Havoc, Properties as Props, To } from "../../singletons.js";
 import type { BabelNodeExpression } from "babel-types";
 import * as t from "babel-types";
 import invariant from "../../invariant.js";
+import type { VisitEntryCallbacks } from "../../utils/generator.js";
+
+// This function attempts to optimize Object.assign calls, by merging mulitple
+// calls into one another where possible. For example:
+//
+// var a = Object.assign({}, someAbstact);
+// var b = Object.assign({}, a);
+//
+// Becomes:
+// var b = Object.assign({}, someAbstract, a);
+//
+// This is a recursive function, so it will attempt to do this multiple times
+// until it can no longer do so, or if a particular Object.assign is visited
+// and thus it is not possible to merge the Object.assign calls together.
+export function attemptToMergeEquivalentObjectAssigns(callbacks: VisitEntryCallbacks, value: Value): boolean {
+  if (!(value instanceof AbstractObjectValue)) {
+    return false;
+  }
+  let realm = value.$Realm;
+  let temporalBuildNodeEntry = realm.getTemporalBuildNodeEntryArgsFromDerivedValue(value);
+  if (temporalBuildNodeEntry === undefined) {
+    return false;
+  }
+  let args = temporalBuildNodeEntry.args;
+  // If we are Object.assigning 3 or more args
+  if (args.length < 3) {
+    return false;
+  }
+  // Then scan through the args after the "to" of this Object.assign, to see if any
+  // other sources are the "to" of a previous Object.assign call
+  for (let i = 2; i < args.length; i++) {
+    let possibleOtherObjectAssignTo = args[i];
+    // Ensure that the "to" value can be omitted
+    if (!callbacks.canOmit(possibleOtherObjectAssignTo)) {
+      continue;
+    }
+    // Check if the "to" was definitely an Object.assign, it should
+    // be a snapshot AbstractObjectValue
+    if (
+      possibleOtherObjectAssignTo instanceof AbstractObjectValue &&
+      // We don't handle conditions for now, this can be done at a later stage
+      possibleOtherObjectAssignTo.kind !== "conditional"
+    ) {
+      let otherTemporalBuildNodeEntry = realm.getTemporalBuildNodeEntryArgsFromDerivedValue(
+        possibleOtherObjectAssignTo
+      );
+      if (otherTemporalBuildNodeEntry === undefined) {
+        continue;
+      }
+      let otherArgs = otherTemporalBuildNodeEntry.args;
+      let objectAssingFunc = realm.intrinsics.Object.$Get("assign", realm.intrinsics.Object);
```

Comment:
I think calling $Get this way might run a getter. Not a good idea in the visitor to run code.

Two thoughts:
1) The fact that the `Object.assign` function is ended in the first argument is a bit silly to begin with. It might cause additional code generation. Don't do that, instead, in the build node, get the `Object.assign` function via the prelude generator's `memoizeReference`, which already happens for special handling of abstract value with `kind === "explicit conversion to object"`. Btw, that other special abstract value is probably also something you'd want to include in your optimization.
2. This way of digging for identifying information, and the whole new custom-optimization apparatus, tells me that this might call for a new sub-class of `GeneratorEntry`, e.g. a `ObjectAssignGeneratorEntry`.